src/main/scala/org/apache/spark/sql/pulsar/PulsarMetadataReader.scala

```diff
@@ -93,30 +102,47 @@ private[pulsar] case class PulsarMetadataReader(
           case _ => throw new RuntimeException(s"Invalid starting time for $tp: $time")
         }
 
+        val (subscriptionNames, subscriptionPredefined) = extractSubscription(subscription, tp)
+
         // setup the subscription
-        try {
-          admin.topics().createSubscription(tp, s"$driverGroupIdPrefix-$tp", msgID)
-        } catch {
-          case _: PulsarAdminException.ConflictException =>
-            log.info("subscription already exists, resetting the cursor to given offset")
-            time match {
-              case PulsarProvider.EARLIEST_TIME | PulsarProvider.LATEST_TIME =>
-                admin.topics().resetCursor(tp, s"$driverGroupIdPrefix-$tp", msgID)
-              case _ =>
-                admin.topics().resetCursor(tp, s"$driverGroupIdPrefix-$tp", time)
-            }
-          case e: Throwable =>
-            throw new RuntimeException(
-              s"Failed to setup cursor for ${TopicName.get(tp).toString}", e)
+        if (!subscriptionPredefined) {
+          try {
+            admin.topics().createSubscription(tp, s"$subscriptionNames", msgID)
+          } catch {
+            case _: PulsarAdminException.ConflictException =>
+              // if subscription already exists, log the info and continue to reset cursor
+              log.info("subscription already exists...")
+            case e: Throwable =>
+              throw new RuntimeException(
+                s"Failed to setup cursor for ${TopicName.get(tp).toString}", e)
+          }
+        }
+
+        // reset cursor position
+        log.info(s"Resetting cursor for $subscriptionNames to given offset")
```

Comment:
Shall we use "timestamp" instead of "offset" here? As offset sounds more close to messageId to me.

Reply:
yeah, I should update this one